<?xml version="1.0"?>
<robot name="snake_12" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- properties -->
  <!-- 1.57079632679 -->
  <xacro:property name="cyl_radius" value="0.0475" />
  <xacro:property name="cyl_diameter" value="0.095" />
  <xacro:property name="cyl_height" value="0.12" />
  <xacro:property name="cyl_mass" value="0.125" /> <!-- in kg-->
  <xacro:property name="rev_axisY" value="0 1 0" />
  <xacro:property name="rev_axisZ" value="0 0 1" />
  <xacro:property name="deg2rad" value="0.01745329251994329577"/>
  <xacro:property name="Deg90Rad" value="1.57079632679"/>

  <macro name="cylinder_inertia" params="mass">
	<!--<origin xyz="0 0 ${cyl_radius}" rpy="0 0 0"/>-->
  <origin xyz="0 0 0" rpy="0 0 0"/>
    <inertia ixx="0.0741666666667" ixy="0.0" ixz="0.0" iyy="0.0585416666667" iyz="0.0" izz="0.0260416666667"/>
  </macro>


  <material name="blue">
    <color rgba="0 0 0.8 1"/>
  </material>
  <material name="black">
    <color rgba="0 0 0 1"/>
  </material>
  <material name="white">
    <color rgba="1 1 1 1"/>
  </material>
  <material name="red">
    <color rgba="1 0 0 1"/>
  </material>

  <xacro:macro name="transmission_block" params="joint_name">
	  <transmission name="tran1">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="${joint_name}">
	      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="motor_${joint_name}">
	      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	      <mechanicalReduction>1</mechanicalReduction>
	    </actuator>
	  </transmission>
   </xacro:macro>


<xacro:macro name="cyl" params="num material">
     <!-- params : number color revAxis { Y , Z } -->
    <link name="link_${num}">
      <visual>

        <!--<origin rpy="0.0 0.0 -${Deg90Rad}" xyz="0.024 ${cyl_radius} ${-cyl_radius}"/>-->
        <origin rpy="0.0 0.0 0.0" xyz="${-cyl_radius} 0.024 ${-cyl_radius}"/>
        <geometry>
          <!--<cylinder length="${cyl_height}" radius="${cyl_radius}" />-->
          <mesh filename="package://robot_snake_simulation/meshes/link_2.STL"/>
        </geometry>
        <material name="${material}" />
      </visual>

      <collision>
        <!--<origin rpy="0.0 0.0 -${Deg90Rad}" xyz="0.024 ${cyl_radius} ${-cyl_radius}"/>-->
        <origin rpy="0.0 0.0 0.0" xyz="${-cyl_radius} 0.024 ${-cyl_radius}"/>
        <geometry>
          <mesh filename="package://robot_snake_simulation/meshes/link_2.STL"/>
        </geometry>
      </collision>
      <inertial>
        <mass value="${cyl_mass}" />
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
        <!--<cylinder_inertia  mass="${cyl_mass}" />-->
      </inertial>
    </link>
  </xacro:macro>

<xacro:macro name="link_fake" params="num">
  <link name="link_fake_${num}">
    <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <sphere radius="0.05"/>
        </geometry>
    </visual>
    <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <sphere radius="0.05"/>
        </geometry>
    </collision>
  </link>
</xacro:macro>

 <xacro:macro name="attachjoint" params="parent child rev">
  <joint name="joint_${child}" type="continuous">
    <parent link="link_${parent}"/>
    <child link="link_${child}"/>
    <origin xyz="0.0 0.16 0.0" rpy="0.0 0.0 0.0" />
    <axis xyz="${rev}" />
    <dynamics damping="0.7"/>
  </joint>
 </xacro:macro>


<link name="world"/>

<link name="base_link" >
  <visual>
    <origin rpy="1.57079632679 0 3.14159265359" xyz="1 -0.5 -0.36"/>
    <geometry>
        <!--<box size="0.1 0.6 0.6"/>-->
      <mesh filename="package://robot_snake_simulation/meshes/base.STL"/>
    </geometry>
  </visual>
  <collision>
    <geometry>
      <box size="2 1 0.72"/>
    </geometry>
  </collision>
  <inertial>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <mass value="100000" />
    <inertia ixx="100000" ixy="0.0" ixz="0.0" iyy="100000" iyz="0.0" izz="100000"/>
  </inertial>
</link>

    <link name="arc" >
      <visual>
        <origin rpy="0 0 0" xyz="-0.6 -0.195 -0.3175"/><!---0.3175-->
        <geometry>
        <!--<box size="0.3 0.3 0.3"/>-->
          <mesh filename="package://robot_snake_simulation/meshes/arc.STL"/>
        </geometry>
      </visual>
    <collision>
    <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="1.2 0.3 0.3"/><!--0.495-->
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.2" />
      <inertia ixx="0.0741666666667" ixy="0.0" ixz="0.0" iyy="0.0585416666667" iyz="0.0" izz="0.0260416666667"/>
   </inertial>
  </link>


  <xacro:cyl num="1"  material="blue" />
  <xacro:cyl num="2"  material="white" />
  <xacro:cyl num="3"  material="blue" />
  <xacro:cyl num="4"  material="white"/>
  <xacro:cyl num="5"  material="blue" />
  <xacro:cyl num="6"  material="white" />
  <xacro:cyl num="7"  material="blue" />
  <xacro:cyl num="8"  material="white" />
  <xacro:cyl num="9"  material="blue" />
  <xacro:cyl num="10" material="white" />
  <xacro:cyl num="11" material="white" />
  <xacro:cyl num="12" material="white" />
  <xacro:link_fake num="1" />
  <xacro:link_fake num="2" />
  <xacro:link_fake num="3" />
  <xacro:link_fake num="4" />
  <xacro:link_fake num="5" />
  <xacro:link_fake num="6" />
  <xacro:link_fake num="7" />
  <xacro:link_fake num="8" />
  <xacro:link_fake num="9" />
  <xacro:link_fake num="10" />
  <xacro:link_fake num="11" />
  <xacro:link_fake num="12" />


  <joint name="fixed" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
    <origin xyz="0.0 0.0 0.5" rpy="0 0 0" />
  </joint>

  <joint name="linear1" type="prismatic">
    <parent link="base_link"/>
    <child link="arc"/>
    <limit effort="0" lower="0" upper="2" velocity="0.5"/>
    <axis xyz="1 0 0" />
    <origin xyz="-0.5 0 ${0.7365-0.03175}" rpy="0 0 0" />
        <dynamics damping="0.7"/>
  </joint>

  <joint name="joint_fake_1" type="continuous">
    <parent link="arc"/>
    <child link="link_fake_1"/>
    <!--<origin xyz="${0.6+0.048} 0 0" rpy="0 0 0" />--><!--0.0525 -->
    <origin xyz="0.648 0.0 ${cyl_radius}" rpy="0.0 0.0 ${-Deg90Rad}" /><!--0.0525 -->
    <axis xyz="0 0 1" />
    <dynamics damping="0.7"/>
  </joint>


<xacro:attachjoint parent="fake_1"  child="1" rev="0 1 0" />

<xacro:attachjoint parent="1"  child="fake_2" rev="0 0 1" />
<xacro:attachjoint parent="fake_2"  child="2" rev="0 1 0" />

<xacro:attachjoint parent="2"  child="fake_3" rev="0 0 1" />
<xacro:attachjoint parent="fake_3"  child="3" rev="0 1 0" />

<xacro:attachjoint parent="3"  child="fake_4" rev="0 0 1" />
<xacro:attachjoint parent="fake_4"  child="4" rev="0 1 0" />

<xacro:attachjoint parent="4"  child="fake_5" rev="0 0 1" />
<xacro:attachjoint parent="fake_5"  child="5" rev="0 1 0" />

<xacro:attachjoint parent="5"  child="fake_6" rev="0 0 1" />
<xacro:attachjoint parent="fake_6"  child="6" rev="0 1 0" />

<xacro:attachjoint parent="6"  child="fake_7" rev="0 0 1" />
<xacro:attachjoint parent="fake_7"  child="7" rev="0 1 0" />

<xacro:attachjoint parent="7"  child="fake_8" rev="0 0 1" />
<xacro:attachjoint parent="fake_8"  child="8" rev="0 1 0" />

<xacro:attachjoint parent="8"  child="fake_9" rev="0 0 1" />
<xacro:attachjoint parent="fake_9"  child="9" rev="0 1 0" />

<xacro:attachjoint parent="9"  child="fake_10" rev="0 0 1" />
<xacro:attachjoint parent="fake_10"  child="10" rev="0 1 0" />

<xacro:attachjoint parent="10"  child="fake_11" rev="0 0 1" />
<xacro:attachjoint parent="fake_11"  child="11" rev="0 1 0" />

<xacro:attachjoint parent="11"  child="fake_12" rev="0 0 1" />
<xacro:attachjoint parent="fake_12"  child="12" rev="0 1 0" />



<gazebo reference="base_link">
  <material>Gazebo/Gray</material>
</gazebo>
  <gazebo reference="arc">
<material>Gazebo/Orange</material>
</gazebo>
  <gazebo reference="link_1">
<material>Gazebo/White</material>
</gazebo>
  <gazebo reference="link_2">
<material>Gazebo/Blue</material>
</gazebo>
<gazebo reference="link_3">
  <material>Gazebo/White</material>
</gazebo>
<gazebo reference="link_4">
  <material>Gazebo/Blue</material>
</gazebo>
<gazebo reference="link_5">
  <material>Gazebo/White</material>
</gazebo>
<gazebo reference="link_6">
  <material>Gazebo/Blue</material>
</gazebo>
<gazebo reference="link_7">
  <material>Gazebo/White</material>
</gazebo>
<gazebo reference="link_8">
  <material>Gazebo/Blue</material>
</gazebo>
<gazebo reference="link_9">
  <material>Gazebo/White</material>
</gazebo>
<gazebo reference="link_10">
  <material>Gazebo/Blue</material>
</gazebo>
<gazebo reference="link_11">
  <material>Gazebo/White</material>
</gazebo>
<gazebo reference="link_12">
  <material>Gazebo/Blue</material>
</gazebo>


<!-- ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////// -->
  <!-- Transmissions for ROS Control -->
<transmission name="linear1">
  <type>transmission_interface/SimpleTransmission</type>
  <joint name="linear1">
    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
  </joint>
  <actuator name="motor_0">
	  <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	  <mechanicalReduction>1</mechanicalReduction>
	</actuator>
</transmission>

   <xacro:transmission_block joint_name="joint_fake_1"/>
   <xacro:transmission_block joint_name="joint_1"/>

   <xacro:transmission_block joint_name="joint_fake_2"/>
   <xacro:transmission_block joint_name="joint_2"/>

   <xacro:transmission_block joint_name="joint_fake_3"/>
   <xacro:transmission_block joint_name="joint_3"/>

   <xacro:transmission_block joint_name="joint_fake_4"/>
   <xacro:transmission_block joint_name="joint_4"/>

   <xacro:transmission_block joint_name="joint_fake_5"/>
   <xacro:transmission_block joint_name="joint_5"/>

   <xacro:transmission_block joint_name="joint_fake_6"/>
   <xacro:transmission_block joint_name="joint_6"/>

   <xacro:transmission_block joint_name="joint_fake_7"/>
   <xacro:transmission_block joint_name="joint_7"/>

   <xacro:transmission_block joint_name="joint_fake_8"/>
   <xacro:transmission_block joint_name="joint_8"/>

   <xacro:transmission_block joint_name="joint_fake_9"/>
   <xacro:transmission_block joint_name="joint_9"/>

   <xacro:transmission_block joint_name="joint_fake_10"/>
   <xacro:transmission_block joint_name="joint_10"/>

   <xacro:transmission_block joint_name="joint_fake_11"/>
   <xacro:transmission_block joint_name="joint_11"/>

   <xacro:transmission_block joint_name="joint_fake_12"/>
   <xacro:transmission_block joint_name="joint_12"/>


  <!-- ros_control plugin -->
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/snake_12</robotNamespace>
      <legacyModeNS>true</legacyModeNS>
    </plugin>
  </gazebo>
<!--
    <xacro:include filename="$(find mastering_ros_robot_description_pkg)/urdf/sensors/xtion_pro_live.urdf.xacro"/>
    -->
  <!-- Define RGB-D sensor -->
  <!--
  <xacro:xtion_pro_live name="rgbd_camera" parent="link10">
   <origin xyz="0.2 0 0" rpy="0 0 0"/>
    <origin xyz="0 0 0" rpy="${-90.0 * deg2rad} 0 ${-90.0 * deg2rad}"/>
  </xacro:xtion_pro_live>
  -->
</robot>
